### PR TITLE
Calculate unhealthy reason (input/output/other) on agent documents

### PR DESCRIPTION
**Pull request by juliaElastic**
elasticsearch pr to add `unhealthy_reason` keyword mapping: https://github.com/elastic/elasticsearch/pull/106246

## What is the problem this PR solves?

Add `unhealthy_reason` to fleet server metrics published regularly.

## How does this PR solve the problem?

Calculate `unhealthy_reason` from `agent.components` on checkin and save in agent doc.

## How to test this PR locally

- enroll an agent with docker
- add endpoint integration, expect an input and output unit error status on the agent doc
- check that `unhealthy_reason` is added to the agent doc

```
GET .fleet-agents/_search
{
  "_source": ["unhealthy_reason"], 
    "sort": [
    {
      "last_checkin": {
        "order": "desc"
      }
    }
  ]
}

   "hits": [
      {
        "_index": ".fleet-agents-7",
        "_id": "f7406324-ed6b-4605-b9c2-05be026332b3",
        "_score": null,
        "_source": {
          "unhealthy_reason": [
            "input", "output"
          ]
        },
        "sort": [
          1710250909000
        ]
      },
```

## Design Checklist

- [ ] I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.
- [ ] I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.
- [ ] I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc.

## Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/fleet-server#changelog)

## Related issues

Relates https://github.com/elastic/ingest-dev/issues/2522

### CODE REVIEW

File: internal/pkg/api/handleCheckin.go [hunk 1]
```diff
@@ -969,9 +979,67 @@ func parseComponents(zlog zerolog.Logger, agent *model.Agent, req *CheckinReques
...
+
+	return outComponents, &unhealthyReason, nil
+}
+
+func calcUnhealthyReason(reqComponents interface{}) ([]string, error) {
+	var unhealthyReason []string
+	hasUnhealthyInput := false
+	hasUnhealthyOutput := false
+	hasUnhealthyComponent := false
+	reqComponentsArray, ok := reqComponents.([]interface{})
```
[the remark this answers is not in the document]
Comment (juliaElastic) [reply internal/pkg/api/handleCheckin.go]: Added the schema and it simplifies the code by a lot.
It needs more testing though, I'm not sure what happens if there are other (unmapped) properties in components, we should probably keep them for debug purposes.

[the remark this answers is not in the document]
Comment (juliaElastic) [reply internal/pkg/api/handleCheckin.go]: tested with a healthy endpoint input, that extra properties are not being removed after adding the schema:
```
 "components": [
    {
      "id": "endpoint-default",
      "type": "endpoint",
      "status": "HEALTHY",
      "message": "Healthy: communicating with endpoint service",
      "units": [
        {
          "id": "endpoint-default",
          "type": "output",
          "status": "HEALTHY",
          "message": "Applied policy {e14510ab-83b8-4c40-af40-519ea5203adf}",
          "payload": {
            "error": {
              "code": 0,
              "message": "Success"
            }
          }
        },
```

**Comment by juliaElastic** [comment]
buildkite run fleet-server-perf-tests

**Comment by juliaElastic** [comment]
buildkite run perf-tests

File: internal/pkg/api/handleCheckin.go [hunk 2]
```diff
@@ -917,50 +922,50 @@ func parseMeta(zlog zerolog.Logger, agent *model.Agent, req *CheckinRequest) ([]
...
-			return nil, fmt.Errorf("parseComponents local: %w", err)
-		}
+		return nil, &unhealthyReason, nil
 	}
 
 	var outComponents []byte
 
 	// Compare the deserialized meta structures and return the bytes to update if different
-	if !reflect.DeepEqual(reqComponents, agentComponents) {
+	if !reflect.DeepEqual(reqComponents, agent.Components) {
```
Comment (nchaulet) [inline internal/pkg/api/handleCheckin.go:965]: Do we have already a test that test we do not update if not different?

Comment (juliaElastic) [reply internal/pkg/api/handleCheckin.go:967]: added tests to verify when components equals / not equals

**Review by nchaulet — APPROVED** [review]
LGTM 🚀

**Pull request merged**